torchtune/dev/grpo/rewards.py

```diff
@@ -5,17 +5,148 @@
 # LICENSE file in the root directory of this source tree.
 
 import re
-from typing import List, Tuple
+from typing import Dict, List, Optional, Protocol, Tuple, Union
 from xml.etree import ElementTree as ET
 
 import torch
-
-# from tensordict import TensorDict, TensorDictBase
-# from torchrl.data import Composite, TensorSpec, Unbounded
-# from torchrl.envs import Transform
+from dataclasses import dataclass, field
 
 from torchtune.modules.transforms.tokenizers import ModelTokenizer
 
+@dataclass
+class RewardOutput:
+    reward_base_name: str
+    total_reward: torch.Tensor
+    successes: torch.Tensor
+    # optional
+    rewards: Optional[Dict[str, torch.Tensor]] = field(default_factory=dict)
+
+    def log(self, prefix: str = "") -> Dict[str, float]:
+        log_dict = {}
+        prefix = f"{prefix}/{self.reward_base_name}" if prefix else self.reward_base_name
+        for reward_name, reward in self.rewards.items():
+            log_dict[f"{prefix}/{reward_name}"] = reward.mean().item()
+        log_dict[f"{prefix}"] = self.total_reward.mean().item()
+        log_dict[f"{prefix}/successes"] = self.successes.mean().item()
+        return log_dict
+
+
+class RewardBase(Protocol):
+    def __call__(
+        self,
+        completion_ids: torch.Tensor,
+        completions: List[str],
+        answers: List[str],
+        device: torch.device,
+    ) -> RewardOutput:
+        pass
+
+
+class FormattedMathCorrectnessReward(RewardBase):
+    """
+    This reward encourages the model to correctly answer a math problem, and requires
+    the model to repond in an XML-style format to extract answers. 
+
+    Args:
+        answer_tag: the tag for the answer section. The answer will be extracted from <{answer_tag}>{answer}</{answer_tag}>
+        positive_reward: the reward provided for correctly formatted completions
+        negative_reward: the reward provided for incorrectly formatted completions
+    """
+    def __init__(self, answer_tag: str, positive_reward: float, negative_reward: float = 0.0):
+        self.answer_tag = answer_tag
+        self.positive_reward = positive_reward
+        self.negative_reward = negative_reward
+
+    def __call__(
+        self,
+        completion_ids: torch.Tensor,
+        completions: List[str],
+        answers: List[str],
+        device: torch.device,
+    ) -> RewardOutput:
+        rewards = []
+        answer_pattern = rf"<{self.answer_tag}>(.*?)</{self.answer_tag}>"
+        for completion, answer in zip(completions, answers):
+            match = re.search(answer_pattern, completion, re.S)
+            if match:
+                answer = match.group(1).strip()
+                reward = self.positive_reward if answer == answer else self.negative_reward
+                rewards.append(reward)
+            else:
+                rewards.append(self.negative_reward)
+            
+        rewards = torch.tensor(rewards)
```

Comment:
Do we not want to put this tensor on the given device, or is that not what the device is for?

Reply:
Yeah, I think for reward functions that use a model, we might want to let the orchestration system place them, and then they can init themselves given the resources they can see locally. Definitely good to leave for followup